File: 110. Balanced Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        height, isBalanced = self.postOrder(root, 0, True)
        return isBalanced
    
    def postOrder(self, current, height, isBalanced):
        if not current:
            return 0, True
        if not isBalanced:
            return height, isBalanced
        leftheight, isBalanced = self.postOrder(current.left, 0, isBalanced)
        if isBalanced:
            rightheight, isBalanced = self.postOrder(current.right, 0, isBalanced)
            if isBalanced:
                return max(leftheight, rightheight) + 1, abs(leftheight - rightheight) < 2
        return height, isBalanced


         
        

File: 110. Balanced Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public boolean isBalanced(TreeNode root) {
        return helper(root).isBalanced;
    }

    public Result helper(TreeNode node){
            if(node == null) {
                return new Result(true, 0);
            }
            Result left = helper(node.left);
            if (!left.isBalanced){
                return new Result(false, 0);
            }
            Result right = helper(node.right);
            if (!right.isBalanced){
                return new Result(false, 0);
            }
            
            int diff =  Math.abs(left.height - right.height);
            if (diff > 1){
                return new Result(false, 0);
            } 
            return new Result(true, 1 + Math.max(left.height, right.height)); 
        }
}

class Result {
    boolean isBalanced = true;
    int height = 0;

    Result(boolean isBalanced, int height){
        this.isBalanced = isBalanced;
        this.height = height;
    }

}
